nexus/db-queries/src/db/queries/regions_hard_delete.rs

```diff
@@ -0,0 +1,60 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of query to update crucible_dataset size_used after
+//! hard-deleting regions
+
+use crate::db::datastore::RunnableQueryNoReturn;
+use crate::db::raw_query_builder::QueryBuilder;
+use diesel::sql_types;
+use uuid::Uuid;
+
+/// Update the affected Crucible dataset rows after hard-deleting regions
+pub fn dataset_update_query(
+    dataset_ids: Vec<Uuid>,
+) -> impl RunnableQueryNoReturn {
+    let mut builder = QueryBuilder::new();
+
+    builder.sql(
+        "WITH
+  size_used_with_reservation AS (
+    SELECT
+      crucible_dataset.id AS crucible_dataset_id,
+      SUM(
+        CASE
+          WHEN block_size IS NULL THEN 0
+          ELSE
+            CASE
+              WHEN reservation_percent = '25' THEN
+                (block_size * blocks_per_extent * extent_count) / 4 +
+                (block_size * blocks_per_extent * extent_count)
+            END
+        END
+      ) AS reserved_size
+    FROM crucible_dataset
+    LEFT JOIN region ON crucible_dataset.id = region.dataset_id
+    WHERE
+      crucible_dataset.time_deleted IS NULL AND
+      crucible_dataset.id IN (",
+    );
+
+    for (idx, dataset_id) in dataset_ids.into_iter().enumerate() {
+        if idx != 0 {
+            builder.sql(",");
+        }
+        builder.param().bind::<sql_types::Uuid, _>(dataset_id);
+    }
+
+    builder.sql(
```

Comment:
Would it be possible to use:

```rust
builder.param().bind::<diesel::pg::sql_types::Array<sql_types::Uuid>, _>(dataset_ids)
```

Instead of creating a new bind parameter for each individual index?

Reply:
I started with this, but hit a runtime error, something like `couldn't map from uuid[] -> uuid` - maybe I was doing it wrong? idk

Reply:
```
thread 'db::queries::regions_hard_delete::test::explainable' panicked at nexus/db-queries/src/db/queries/regions_hard_delete.rs:104:14:
Failed to explain query - is it valid SQL?: DatabaseError(Unknown, "invalid cast: uuid[] -> uuid")
```

Reply:
The following works for me:

```
-      crucible_dataset.id IN (",
-    );
-
-    for (idx, dataset_id) in dataset_ids.into_iter().enumerate() {
-        if idx != 0 {
-            builder.sql(",");
-        }
-        builder.param().bind::<sql_types::Uuid, _>(dataset_id);
-    }
-
-    builder.sql(
-        ")
+      crucible_dataset.id = ANY (",
+        )
+        .param()
+        .bind::<diesel::pg::sql_types::Array<sql_types::Uuid>, _>(dataset_ids)
+        .sql(
+            ")
```

Reply:
Nice, yeah - I missed how `IN` is not the same as `= ANY`, changed in fc3735825